self
error
failed_record
